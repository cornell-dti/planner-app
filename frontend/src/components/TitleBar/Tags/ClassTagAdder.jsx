// @flow strict

import React from 'react';
import type { ComponentType } from 'react';
import ReactSearchBox from 'react-search-box';
import { connect } from 'react-redux';
import styles from './TagAdder.css';
import type { Course } from '../../../store/store-types';
import { addTag } from '../../../firebase/actions';

type Props = {| +courses: Map<number, Course[]>; |};

type SimpleCourse = {|
  +key: number;
  +value: string;
  +subject: string;
  +courseNumber: string;
  +title: string;
  +classId: number;
|};

/**
 * Returns the computed course options.
 *
 * @param {Map<number, Course[]>} courseMap new courses.
 * @return {SimpleCourse[]} course options.
 */
function getCourseOptions(courseMap: Map<number, Course[]>): SimpleCourse[] {
  const courseOptions = [];
  let i = 0;
  courseMap.forEach((courses: Course[]) => {
    courses.forEach((course: Course) => {
      const {
        subject, courseNumber, title, courseId: classId,
      } = course;
      const name = `${subject} ${courseNumber}: ${title}`;
      courseOptions.push({
        key: i, value: name, subject, courseNumber, title, classId,
      });
      i += 1;
    });
  });
  return courseOptions;
}

/**
 * The configs for fuse searcher. Essential for fuzzy search.
 * You may need to tune this further, but it's usable right now.
 *
 * @type {{keys: string[], threshold: number}}
 */
const fuseConfigs = {
  keys: [
    'value', // useful for search the full name as we display
    'subject', // useful for finding a list of all [subject] classes
    'courseNumber', // useful if the student just type the course number
    'title', // useful if the student just type the name
  ],
  location: 0, // since we have customized the stuff to search, we can just start at beginning.
  threshold: 0.2, // higher the threshold, more stuff will be matched.
};

/**
 * The class tag adder.
 *
 * @param {Map<number, Course[]>} courses all courses.
 * @return {Node} the rendered node.
 * @constructor
 */
function ClassTagAdder({ courses }: Props) {
  if (courses.size === 0) {
    return null;
  }
  const changeClass = (option: SimpleCourse) => {
    const { value, classId } = option;
    addTag({
      name: value, color: '#56d9c1', classId,
    });
  };
  return (
    <div className={`${styles.TagColorConfigItemAdder} ${styles.SearchClasses}`}>
      <ReactSearchBox
        data={getCourseOptions(courses)}
        value=""
        fuseConfigs={fuseConfigs}
        onSelect={changeClass}
        placeholder="Search for classes (e.g. CS 2110, Introduction to Creative Writing)"
      />
    </div>
  );
}

const ConnectedClassTagAdder: ComponentType<{||}> = connect(
  ({ courses }) => ({ courses }), null,
)(React.memo(ClassTagAdder));
export default ConnectedClassTagAdder;
