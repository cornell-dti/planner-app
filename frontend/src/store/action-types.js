// @flow strict

import type {
  PartialMainTask, PartialSubTask, SubTask, Tag, Task, Course,
} from './store-types';
import type { TaskDiff } from '../util/task-util';

/*
 * --------------------------------------------------------------------------------
 * Part 1: Task Actions
 * --------------------------------------------------------------------------------
 */

export type AddNewTaskAction = {| type: 'ADD_NEW_TASK'; +task: Task; |};
export type AddNewSubTaskAction = {|
  type: 'ADD_NEW_SUBTASK'; +taskId: number; +subTask: SubTask;
|};

export type EditTaskAction = {| type: 'EDIT_TASK'; +task: Task; +diff: TaskDiff; |};
export type EditMainTaskAction = {|
  type: 'EDIT_MAIN_TASK'; +taskId: number; +partialMainTask: PartialMainTask;
|};
export type EditSubTaskAction = {|
  type: 'EDIT_SUB_TASK'; +taskId: number; +subtaskId: number; +partialSubTask: PartialSubTask;
|};

export type RemoveTaskAction = {| type: 'REMOVE_TASK'; +taskId: number; |};
export type RemoveSubTaskAction = {|
  type: 'REMOVE_SUBTASK'; +taskId: number; +subtaskId: number;
|};

export type ImportCourseExamsAction = {| type: 'IMPORT_COURSE_EXAMS' |};

type TaskAction =
  | AddNewTaskAction
  | AddNewSubTaskAction
  | EditTaskAction
  | EditMainTaskAction
  | EditSubTaskAction
  | RemoveTaskAction
  | RemoveSubTaskAction
  | ImportCourseExamsAction;

/*
 * --------------------------------------------------------------------------------
 * Part 2: Color Config Actions
 * --------------------------------------------------------------------------------
 */

export type AddTagAction = {| type: 'ADD_TAG'; +tag: Tag; |};
export type EditTagAction = {| type: 'EDIT_TAG'; +tag: Tag; |};
export type RemoveTagAction = {| type: 'REMOVE_TAG'; +tagId: number; |};

export type TagAction = AddTagAction | EditTagAction | RemoveTagAction;

/*
 * --------------------------------------------------------------------------------
 * Part 3: Undo Actions
 * --------------------------------------------------------------------------------
 */

export type UndoAddTaskAction = {| type: 'UNDO_ADD_TASK' |};
export type ClearUndoAddTaskAction = {| type: 'CLEAR_UNDO_ADD_TASK' |};
export type UndoDeleteTaskAction = {| type: 'UNDO_DELETE_TASK' |};
export type ClearUndoDeleteTaskAction = {| type: 'CLEAR_UNDO_DELETE_TASK' |};

export type UndoAction =
  | UndoAddTaskAction
  | ClearUndoAddTaskAction
  | UndoDeleteTaskAction
  | ClearUndoDeleteTaskAction;

/*
 * --------------------------------------------------------------------------------
 * Part 4: Backend Patch Actions
 * --------------------------------------------------------------------------------
 */

export type BackendPatchNewTagAction = {|
  type: 'BACKEND_PATCH_NEW_TAG'; +tempId: number; +backendId: number;
|};
export type BackendPatchNewTaskAction = {|
  type: 'BACKEND_PATCH_NEW_TASK'; +tempId: number; +backendTask: Task;
|};
export type BackendPatchBatchNewTasksAction = {|
  type: 'BACKEND_PATCH_BATCH_NEW_TASKS'; +tempIds: number[]; +backendTasks: Task[];
|};
export type BackendPatchNewSubTaskAction = {|
  type: 'BACKEND_PATCH_NEW_SUBTASK';
  +taskId: number;
  +tempSubTaskId: number;
  +backendSubTaskId: number;
|};
export type BackendPatchExistingTaskAction = {|
  type: 'BACKEND_PATCH_EXISTING_TASK'; +task: Task;
|};
export type BackendPatchLoadedDataAction = {|
  type: 'BACKEND_PATCH_LOADED_DATA'; +tags: Tag[]; +tasks: Task[]; +courses: Map<number, Course[]>;
|};

/**
 * All the tasks that lets the backend patch the store with all the latest values.
 */
export type BackendPatchAction =
  | BackendPatchNewTagAction
  | BackendPatchNewTaskAction
  | BackendPatchBatchNewTasksAction
  | BackendPatchNewSubTaskAction
  | BackendPatchExistingTaskAction
  | BackendPatchLoadedDataAction

/*
 * --------------------------------------------------------------------------------
 * Part 5: All Actions
 * --------------------------------------------------------------------------------
 */

export type Action = TagAction | TaskAction | UndoAction | BackendPatchAction;
