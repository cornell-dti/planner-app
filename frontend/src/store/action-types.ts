import { Map } from 'immutable';
import { Course, Settings, SubTask, Tag, Task, BannerMessageStatus } from './store-types';
import { type } from 'os';

export type PatchTags = {
  readonly type: 'PATCH_TAGS';
  readonly created: Tag[];
  readonly edited: Tag[];
  readonly deleted: string[];
};

export type PatchTasks = {
  readonly type: 'PATCH_TASKS';
  readonly created: Task[];
  readonly edited: Task[];
  readonly deleted: string[];
};

export type PatchSubTasks = {
  readonly type: 'PATCH_SUBTASKS';
  readonly created: SubTask[];
  readonly edited: SubTask[];
  readonly deleted: string[];
};

export type PatchSettings = {
  readonly type: 'PATCH_SETTINGS';
  readonly settings: Settings;
};

export type PatchBannerMessageStatus = {
  readonly type: 'PATCH_BANNER_MESSAGES';
  readonly change: BannerMessageStatus;
};

export type PatchCourses = {
  readonly type: 'PATCH_COURSES';
  readonly courses: Map<string, Course[]>;
};

export type Action =
  | PatchTags
  | PatchTasks
  | PatchSubTasks
  | PatchSettings
  | PatchBannerMessageStatus
  | PatchCourses;
