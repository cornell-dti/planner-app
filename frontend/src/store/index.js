// @flow strict

import { createStore } from 'redux';
import type { Store } from 'redux';
import rootReducer from './reducers';
import type { State } from './store-types';
import type { Action } from './action-types';

const store: Store<State, $Subtype<Action>> = createStore(rootReducer);
export default store;
