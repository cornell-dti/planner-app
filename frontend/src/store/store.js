// @flow strict

import { createStore } from 'redux';
import type { Store } from 'redux';
import rootReducer from './reducers';
import type { State } from './store-types';
import type { Action } from './action-types';
import type { AppUser } from '../util/firebase-util';
import { error } from '../util/general-util';

export type GlobalStore = Store<State, $Subtype<Action>>;

const store: GlobalStore = createStore(rootReducer);
let appUser: AppUser | null = null;

/**
 * Initialize the app and record a global user and returns the initialized store.
 *
 * @param {AppUser} user the user of the app.
 * @return {GlobalStore} the newly initialized store.
 */
export function initializeApp(user: AppUser): GlobalStore {
  appUser = user;
  return store;
}

/**
 * Returns the global app user.
 *
 * @return {AppUser} the global app user.
 */
export const getAppUser = (): AppUser => appUser ?? error('App is not initialized.');

/**
 * Dispatch an action and returns the dispatched action.
 *
 * @param {Action} action the action to dispatch.
 * @return {Action} the action to dispatch.
 */
export const dispatchAction = (action: Action): Action => store.dispatch(action);
