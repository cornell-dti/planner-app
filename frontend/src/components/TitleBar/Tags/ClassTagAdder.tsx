import React, { useState, ReactElement } from 'react';
import { Map } from 'immutable';
// @ts-ignore we need to rewrite the entire component!
import ReactSearchBox from 'react-search-box';
import { connect } from 'react-redux';
import styles from './TagAdder.module.css';
import { Course, State } from '../../../store/store-types';
import { addTag } from '../../../firebase/actions';
import getUnusedColor from './rotation-color-picker';

type Props = { readonly courses: Map<string, Course[]> };

type SimpleCourse = {
  readonly key: number;
  readonly value: string;
  readonly subject: string;
  readonly courseNumber: string;
  readonly title: string;
  readonly classId: string;
};

/**
 * Returns the computed course options.
 *
 * @param {Map<number, Course[]>} courseMap new courses.
 * @return {SimpleCourse[]} course options.
 */
function getCourseOptions(courseMap: Map<string, Course[]>): SimpleCourse[] {
  const courseOptions: SimpleCourse[] = [];
  let i = 0;
  courseMap.forEach((courses: Course[]) => {
    courses.forEach((course: Course) => {
      const {
        subject, courseNumber, title, courseId: classId,
      } = course;
      const id = `${classId} ${subject} ${courseNumber}`;
      const name = `${subject} ${courseNumber}: ${title}`;
      courseOptions.push({
        key: i, value: name, subject, courseNumber, title, classId: id,
      });
      i += 1;
    });
  });
  return courseOptions;
}

/**
 * The configs for fuse searcher. Essential for fuzzy search.
 * You may need to tune this further, but it's usable right now.
 */
const fuseConfigs: { keys: string[]; location: number; threshold: number } = {
  keys: [
    'value', // useful for search the full name as we display
    'subject', // useful for finding a list of all [subject] classes
    'courseNumber', // useful if the student just type the course number
    'title', // useful if the student just type the name
  ],
  location: 0, // since we have customized the stuff to search, we can just start at beginning.
  threshold: 0.2, // higher the threshold, more stuff will be matched.
};

/**
 * The class tag adder.
 */
function ClassTagAdder({ courses }: Props): ReactElement | null {
  const [key, setKey] = useState(0);
  if (courses.size === 0) {
    return null;
  }
  const changeClass = (option: SimpleCourse): void => {
    const { value, classId } = option;
    addTag({
      name: value, color: getUnusedColor(), classId,
    });
    // force the react search box to rerender due to its bug.
    setKey(prev => prev + 1);
  };
  return (
    <div className={`${styles.TagColorConfigItemAdder} ${styles.SearchClasses}`} title="Search for a class">
      <ReactSearchBox
        tabIndex={0}
        data={getCourseOptions(courses)}
        value=""
        fuseConfigs={fuseConfigs}
        onSelect={changeClass}
        placeholder="Search for classes (e.g. CS 2110, Introduction to Creative Writing)"
        key={key}
      />
    </div>
  );
}

const Memoized = React.memo<Props>(ClassTagAdder);
const Connected = connect(({ courses }: State) => ({ courses }))(Memoized);
export default Connected;
