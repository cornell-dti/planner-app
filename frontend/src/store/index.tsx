import React, { ReactElement } from 'react';
import { Provider as ReactReduxProvider } from 'react-redux';
import { store, storeForTesting } from './store';

type Props = { readonly children: ReactElement };

export const Provider = ({ children }: Props): ReactElement => (
  <ReactReduxProvider store={store}>{children}</ReactReduxProvider>
);

export const ProviderForTesting = ({ children }: Props): ReactElement => (
  <ReactReduxProvider store={storeForTesting}>{children}</ReactReduxProvider>
);
