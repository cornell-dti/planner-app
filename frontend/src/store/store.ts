import { createStore, Store } from 'redux';
import rootReducer from './reducers';
import { State } from './store-types';
import { Action } from './action-types';
import { initialStateForTesting } from './state';

export type GlobalStore = Store<State, Action>;

/**
 * The redux store. It should only be given to the react-redux provider.
 */
export const store: GlobalStore = createStore(rootReducer);

/**
 * The redux store for testing.
 */
export const storeForTesting: GlobalStore = createStore(
  (initialState: State = initialStateForTesting, action: Action): State => rootReducer(
    initialState, action,
  ),
);

/**
 * Dispatch an action and returns the same dispatched action.
 */
export const dispatchAction = (action: Action): Action => store.dispatch(action);
