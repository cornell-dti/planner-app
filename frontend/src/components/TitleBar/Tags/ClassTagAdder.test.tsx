import React from 'react';
import renderer from 'react-test-renderer';
import { ProviderForTesting } from 'store';
import ClassTagAdder from './ClassTagAdder';

it('ClassTagAdder matches snapshot.', () => {
  const tree = renderer.create(<ProviderForTesting><ClassTagAdder /></ProviderForTesting>).toJSON();
  expect(tree).toMatchSnapshot();
});
