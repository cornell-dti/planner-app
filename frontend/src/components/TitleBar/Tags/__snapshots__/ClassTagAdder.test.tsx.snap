// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassTagAdder matches snapshot. 1`] = `
<div
  className="TagColorConfigItemAdder SearchClasses"
  title="Search for a class"
>
  <div>
    <input
      className="SearchInput"
      onChange={[Function]}
      placeholder="Search for classes (e.g. CS 2110, Introduction to Creative Writing)"
      type="text"
      value=""
    />
    <div />
  </div>
</div>
`;
